src/libraries/System.Private.CoreLib/src/System/Collections/Generic/EqualityComparer.cs

```diff
@@ -262,4 +260,278 @@ public void GetObjectData(SerializationInfo info, StreamingContext context)
+
+            if (x is HashSet<T> hashSetx && hashSetx.Comparer.Equals(_elementComparer) &&
+                y is HashSet<T> hashSety && hashSety.Comparer.Equals(_elementComparer))
+                return hashSetx.SetEquals(hashSety);
+
+            // Otherwise, do an O(N^2) match.
+            foreach (T yi in y)
+            {
+                bool found = false;
+                foreach (T xi in x)
+                {
+                    if (_elementComparer.Equals(yi, xi))
+                    {
+                        found = true;
+                        break;
+                    }
+                }
+
+                if (!found)
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public override int GetHashCode([DisallowNull] TSet obj) => obj switch
+        {
+            IReadOnlySet<T> set => set.Count == 0 ? 1 : 2,
+            null => 0
+        };
+
+        // Equals method for the comparer itself.
+        public override bool Equals([NotNullWhen(true)] object? obj) =>
+            obj is not null &&
+            obj is SetEqualityComparer<TSet, T> other &&
+            _elementComparer == other._elementComparer;
+
+        public override int GetHashCode() => HashCode.Combine(
+            GetType().GetHashCode(),
+            _elementComparer.GetHashCode());
+    }
+
+    internal sealed partial class DictionaryEqualityComparer<TDictionary, TKey, TValue> : EqualityComparer<TDictionary>
+        where TDictionary : IReadOnlyDictionary<TKey, TValue>
+    {
+        private readonly IEqualityComparer<TKey> _keyComparer;
+        private readonly IEqualityComparer<TValue> _valueComparer;
+
+        public DictionaryEqualityComparer(
+            IEqualityComparer<TKey>? keyComparer = null,
+            IEqualityComparer<TValue>? valueComparer = null)
+        {
+            _keyComparer = keyComparer ?? EqualityComparer<TKey>.Default;
+            _valueComparer = valueComparer ?? EqualityComparer<TValue>.Default;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public override bool Equals(TDictionary? x, TDictionary? y)
+        {
+            if (x is null)
+                return y is null;
+            if (y is null)
+                return false;
+
+            if (x.Count != y.Count)
+                return false;
+
+            // if (x is Dictionary<TKey, TValue> dictx && dictx.Comparer.Equals(_keyComparer) &&
+            //     y is Dictionary<TKey, TValue> dicty && dicty.Comparer.Equals(_keyComparer))
+            // {
+            //     foreach ((TKey yKey, TValue yValue) in y)
+            //     {
+            //         if (!x.TryGetValue(yKey, out TValue xValue) || !_valueComparer.Equals(yValue, xValue))
+            //             return false;
+            //     }
+            //
+            //     return true;
+            // }
```

Comment:
This optimization cannot be activated here, as the `Dictionary<TKey, TValue>` class has the constraint `where TKey : notnull` which cannot be enforced here.
Is there a way to solve this?